src/software/ai/navigator/path_planner/path_planner_test.cpp

```diff
@@ -0,0 +1,239 @@
+            std::stringstream fail_ss;
+            fail_ss << "Point " << pt << "at s=" << i * PATH_CHECK_INTERVAL
+                    << " is not in navigable area " << navigable_area;
+            throw fail_ss.str();
+        }
+    }
+
+    // check that the path reaches the destination if the destination is not in an
+    // obstacle otherwise check that the path makes progress toward the destination i.e.
+    // the distance to dest is less at the end of the path than at the start of the path
+    bool dest_in_obstacle = std::any_of(
+        obstacles.begin(), obstacles.end(),
+        [&path](const auto &obs) { return obs.containsPoint(path.valueAt(1.0)); });
+
+    if (dest_in_obstacle &&
+        (path.valueAt(1.0) - dest).length() >= (path.valueAt(0.0) - dest).length())
+    {
+        // fail because no progress to destination
+        std::stringstream fail_ss;
+        fail_ss
+            << "Destination is in obstacle, but path does not make progress toward the destination!";
+        throw fail_ss.str();
+    }
+
+    if ((path.valueAt(1.0) - dest).length() <= DEST_CHECK_EPSILON)
+    {
+        // fail because didn't reach destination
+        std::stringstream fail_ss;
+        fail_ss << "Path ends at " << path.valueAt(1.0) << " but dest is " << dest;
+        throw fail_ss.str();
+    }
+    // we passed, yay!
+}
+
+class PlannerTest
+    : public testing::TestWithParam<
+          std::tuple<std::pair<std::string, PathPlannerConstructor>, PlannerTestCase>>
+{
+};
+
+
+TEST_P(PlannerTest, test_path_planner)
+{
+    std::unique_ptr<PathPlanner> planner = std::get<0>(GetParam()).second();
+    auto planner_test_case               = std::get<1>(GetParam());
+    std::optional<Path> path =
+        planner->findPath(planner_test_case.start, planner_test_case.dest,
+                          planner_test_case.navigable_area, planner_test_case.obstacles);
+    if (planner_test_case.should_return_path)
+    {
+        if (path)
+        {
+            try
+            {
+                validatePath(*path, planner_test_case.start, planner_test_case.dest,
+                             planner_test_case.navigable_area,
+                             planner_test_case.obstacles);
+            }
+            catch (const std::string &err)
+            {
+                FAIL() << std::get<0>(GetParam()).first << " failed on case \""
+                       << planner_test_case.name << "\" because " << err;
+            }
+        }
+        else
+        {
+            FAIL() << std::get<0>(GetParam()).first
+                   << "should have returned path for case " << planner_test_case.name;
+        }
+    }
+    else
+    {
+        EXPECT_FALSE(path.has_value())
+            << std::get<0>(GetParam()).first << " should not have returned path for "
+            << planner_test_case.name;
+    }
+}
+
+// uncomment this to run tests
+// cases still fail because theta* will intersect obstacles' bounding boxes at knots
```

Comment:
Can you elaborate on this a little bit / file an issue to fix this?

Reply:
#1207